### ISSUE TITLE
Long-lived LdapConnection causes unconstrained memory growth.

### ISSUE THREAD

**Issue by PaulTrampert** [dsbenghe/Novell.Directory.Ldap.NETStandard#138]
**Issue**
If an LdapConnection is long-lived, doing multiple searches can lead to unconstrained memory growth with no obvious way to fix the issue, as the messages are kept in Connection.cs's `_messages` field. Worked around the issue by casting `ILdapConnection` to `LdapConnection` and `ILdapSearchResults` to `LdapSearchResults` so I could call `connection.Abandon(results)`, but at the very least, this is a very leaky abstraction, if not an outright bug.

**Thoughts on a fix**
This lib could benefit a lot from built in connection pooling that would handle clearing `_messages` when connections are returned to the pool. If no one gets to this before I do, I'll see if I can work out a PR, to implement it, but someone who knows this codebase better than I might be better off doing this.

Quick fix though would be to expose `Abandon` on `ILdapConnection`, and make it take `ILdapSearchResults` rather than `LdapSearchResults` as a parameter.

**Comment by dsbenghe** [comment]
I don't think this classifies as a memory leak - changed the title to better represent the issue.

**Comment by safari137** [comment]
which part of this is added to the milestone?

exposing Abandon and ILdapSearchResults in the interface?
or
connection pooling that does all of this for us?

**Comment by nwoolls** [comment]
@dsbenghe feel free to reject but what about something like PR #271 for handling both pooling and resetting resources?
